medcat/relation_extraction.py

```diff
@@ -0,0 +1,631 @@
+        
+        pos_idxs = numpy.random.choice(pool, size=min(int(self.batch_size//2), len(pool)), replace=False)
+
+        neg_idxs = []
+
+        if numpy.random.uniform() > 0.5:   
+            pool = self.rel_data.relations_dataframe[((self.rel_data.relations_dataframe["ent1"] != ent1_text) | \
+                                                     (self.rel_data.relations_dataframe["ent2"] != ent2_text))].index
+        else:
+            if numpy.random.uniform() > 0.5: # share e1 but not e2
+                pool = self.rel_data.relations_dataframe[(( self.rel_data.relations_dataframe['ent1'] == ent1_text) & \
+                     (self.rel_data.relations_dataframe['ent2'] != ent2_text))].index
+            else: # share e2 but not e1
+                pool = self.rel_data.relations_dataframe[(( self.rel_data.relations_dataframe['ent1'] != ent1_text) & \
+                     (self.rel_data.relations_dataframe['ent2'] == ent2_text))].index
+
+            if len(pool) == 0:
+                    pool = self.rel_data.relations_dataframe[((self.rel_data.relations_dataframe["ent1"] != ent1_text) | \
+                        (self.rel_data.relations_dataframe["ent2"] != ent2_text))].index
+                                                     
+        neg_idxs = numpy.random.choice(pool, size=min(int(self.batch_size//2), len(pool)), replace=False)
+        Q = 1/len(pool)
+
+        logging.debug(" Pos Idx: " + str(pos_idxs))
+        logging.debug(" Neg Idx: " + str(neg_idxs))
+
+        batch = []
+        ## process positive sample
+        pos_df = self.rel_data.relations_dataframe.loc[pos_idxs]
+        for idx, row in pos_df.iterrows():
+            relation, ent1_text, ent2_text = row[0], row[1], row[2]
+            relation_tokens, masked_for_pred, e1_e2_start = self.tokenize(self.put_blanks((relation, ent1_text, ent2_text)))
+            relation_tokens = torch.LongTensor(relation_tokens)
+
+            masked_for_pred = torch.LongTensor(masked_for_pred)
+            e1_e2_start = torch.tensor(e1_e2_start)
+            batch.append((relation_tokens, masked_for_pred, e1_e2_start, torch.FloatTensor([1.0]),\
+                            torch.LongTensor([1])))
+        
+        ## process negative samples
+        negs_df = self.rel_data.relations_dataframe.loc[neg_idxs]
+        for idx, row in negs_df.iterrows():
+            relation, ent1_text, ent2_text = row[0], row[1], row[2]
+            relation_tokens, masked_for_pred, e1_e2_start = self.tokenize(self.put_blanks((relation, ent1_text, ent2_text)))
+            relation_tokens = torch.LongTensor(relation_tokens)
+            masked_for_pred = torch.LongTensor(masked_for_pred)
+            e1_e2_start = torch.tensor(e1_e2_start)
+    
+            batch.append((relation_tokens, masked_for_pred, e1_e2_start, torch.FloatTensor([Q]), torch.LongTensor([0])))
+
+        batch = self.padding_seq(batch)
+
+        return batch
+
+
+class Two_Headed_Loss(nn.Module):
```

Comment:
Why is this neccessary - papers / links to why modelling both is a good loss function?